---
title: "Comparing models of molecular evolution for phylogenetic analysis of the scale worm family Polynoidae"
author: "Jason Moggridge 1159229"
date: "10/12/2020"
bibliography: polynoid_refs.bib
output: 
  pdf_document:
    highlight: kate
    toc: true
urlcolor: blue
---

----

This work is available at [github.com/jmoggridge/6210_project5_phylogeny](https://github.com/jmoggridge/6210_project5_phylogeny)

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = TRUE, message = FALSE, warning = FALSE)
options(tinytex.verbose = TRUE)
```

The following software libraries were used in this work:

```{r installs, eval=FALSE, include=FALSE}
# install.packages('tidyverse')
# install.packages('bold')
# install.packages('ape')
# install.packages('phangorn')
# install.packages('dendextend')
# install.packages('ggdendro')
# install.packages('rcartocolor')
# install.packages('ggthemes')
# install.packages('patchwork')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("Biostrings")
# install.packages("RSQLite")
# BiocManager::install('DECIPHER')
```

```{r libraries}
library(tidyverse) 
library(bold) # data source api
library(Biostrings) # for sequence handling (*clashes w dplyr)
library(ape) # general phylogeny functions, eg. dist.dna
library(DECIPHER) # for alignment, etc.
library(phangorn) # ML phylogeny, model testing
library(dendextend) # dendrograms, tanglegrams
# library(phylogram) # loaded later for dendrogram
library(rcartocolor) # non-ugly discrete color pals
library(ggthemes) # <3 tufte theme for ggplot <3
library(patchwork) # easy plot assembly using +,/,()
```

------

### 1. Introduction

  Phylogenetic analysis is common in biodiversity studies and environmental assessment, however it can be far from clear how to proceed. Not only are there are various phylogenetic marker genes (or combinations thereof) to assess, types of alignment (nucleic acid or protein, symbolic or structural) and methods for estimating phylogeny (distance, maximum-parsimony, maximum-likelihood). In maximum-likelihood, molecular evolution is modeled as a Markov process (*ie*. memoryless) with parameters for the base frequencies and rates of substitution. More complex models account for different substitution rates and base frequencies, *eg*. General time reversible (GTR). Additional models describe variance in rates across the sequence (+G) and proportion of invariant positions (+I).    
    
   Maximum likelihood-based (ML) phylogenetic clustering methods optimize model parameters to maximize the probability of the data given the model [@Felsenstein1981]. As no model is a perfect representation of nature, we need to apply some measure of goodness-of-fit (*eg* log-likelihood, AIC, BIC) to evaluate competing models [@Posada2001; @Posada2008]. Perhaps this model selection step is unnecessary and we can generally use a complex model, GTR+I+G, as we are not particularly interested in the distances, but rather the phylogenetic inferences that we make based on them[@Abadi2019].   
    
   In this work, I sought to explore the impact of model choice in whether model testing is worthwhile. Specifically, I wanted to find out whether different distance models would lead to altered phylogenetic trees. For this, I performed model testing to evaluate AIC, BIC and likelihood criteria, pairwise statistical tests of log-likelihood ratios and tree distances, and a visual comparison of two maximum-likelihood phylogenies in a tanglegram. I chose to use cytochrome c oxidase subunit I sequences of Polynoidae as a model dataset. Polynoidae are a large and speciose family of scale worms that have colonized challenging habitats in the deep sea, including sea caves and hydrothermal vents species[@Zhang2017a]. New species are consistently found in surveys and determining their phylogenetic relationships to known species is of interest[@Hatch2020a].    

-----

### 2. Polynoid DNA sequence dataset acquisition

```{r eval=FALSE}
# Download BOLD specimen + seq data for taxon 'Polynoidae'
polynoids.bold <- bold::bold_seqspec(taxon = 'Polynoidae')
polynoids.bold <- polynoids.bold %>%
  select(processid, contains('name'), contains('marker'), nucleotides) %>%
  select(-c(trace_names, phylum_name, class_name, subspecies_name,
            marker_codes))
write_rds(polynoids.bold, 'polynoid.raw.rds', compress = 'gz')
```
  
Polynoidae sequence data were downloaded from the public Barcode of Life Database (BOLD) api on 2020-12-12, using their `bold` package. There were 1850 specimen records for this taxon in the database in total. The data of interest (identifiers, sequences, and taxonomic names) were saved as `polynoid.raw.rds`.     
  
----

\newpage

### 3. Data quality control and sample selection

Functions for filtering and EDA plot

```{r tidy and filter}
## Selection of data for analysis:
trim_and_filter_seqs <- function(bold.df, markers, minlen,
                                 maxlen, Nthreshold){
  # keep seqs of marker gene; trim N and gaps; 
  # filter by length and Ns proportion
  bold.df <- bold.df %>%
    filter(!is.na(nucleotides) & markercode %in% markers) %>%
    mutate(seq = str_remove_all(nucleotides, '\\s|-|^N+|N+$'),
           seqlen = nchar(seq),
           N.prop = str_count(seq, 'N')/seqlen) %>%
    filter(seqlen >= minlen & seqlen <= maxlen & N.prop <= Nthreshold)
  return(bold.df)  
}
EDA_plot <- function(bold.df, label) {
  # plot sequence lengths, GC%, species per genus
  a <- ggplot(bold.df, aes(x = seqlen)) +
    labs(x = 'Sequence length', subtitle = label) +
    geom_histogram() + geom_rangeframe() + theme_tufte()
  b <- bold.df %>%
    mutate(`% GC` = str_count(seq, '[GC]')/seqlen*100) %>%
    ggplot(aes(x=`% GC`)) + 
    geom_histogram() + geom_rangeframe() + theme_tufte()
  c <-bold.df %>%
    group_by(genus_name) %>%
    summarize(species = length(unique(species_name))) %>%
    ggplot(aes(x=species)) + 
    xlab('Species per genus') +
    geom_histogram() + geom_rangeframe() +theme_tufte()
  return(a|b|c)
}
```

Selection of *cytochrome c oxidase subunit 1* sequences & QC

```{r tidy and eda, fig.height=2, fig.width=7}
# read in data & tidy
polynoids.bold <- read_rds('polynoid.raw.rds') %>%
  mutate(across(where(is.character), ~na_if(.x, ''))) %>%
  mutate(across(where(is.character), as_factor)) %>%
  mutate(nucleotides = as.character(nucleotides)) 
# remove any specimens missing genus & species names
polynoids.bold <- polynoids.bold %>%
  filter(!is.na(genus_name) & !is.na(species_name)) 
# check out all COI sequences (not shown)
coi.eda.all <- polynoids.bold %>%
  trim_and_filter_seqs(markers = 'COI-5P', minlen = 0,
                       maxlen = 1000, Nthreshold = 1) %>%
  EDA_plot('all COI')
summary(str_count(polynoids.bold %>%
                    trim_and_filter_seqs(markers = 'COI-5P', minlen = 500,
                                         maxlen = 800, Nthreshold = 1) %>%
                    pull(seq),'N'))


# apply filtering steps
polynoids.df <- polynoids.bold %>%
  filter(!is.na(genus_name) & !is.na(species_name)) %>%
  trim_and_filter_seqs(markers = 'COI-5P', minlen = 630,
                       maxlen = 680, Nthreshold = 0.05)
coi.eda.filtered <- EDA_plot(polynoids.df, label = 'filtered COI')

# tidy_bold_data() %>%
# Sample 1 sequence per species for analysis, create labels
set.seed(1)
polynoids.sample <- polynoids.df %>%
  group_by(species_name) %>%
  sample_n(1) %>%
  data.frame() %>%
  mutate(seqlabel = paste(genus_name, row_number())) %>%
  mutate(across(where(is.factor), fct_drop)) # drop empty factor levels
# Exploratory analysis plot
coi.sample.eda <- EDA_plot(polynoids.sample, 
         paste('Sampled data: n =', nrow(polynoids.sample)))
```

```{r alignment}
do_alignment <- function(sample.df, names_vector, browse=FALSE, verbose = FALSE){
  # Format and label seqs, reorient reverse complements, align
  seqs <- Biostrings::DNAStringSet(sample.df$seq)
  names(seqs) <- names_vector
  seqs <-  DECIPHER::OrientNucleotides(seqs, verbose = verbose)
  seqs.aligned <- DECIPHER::AlignSeqs(seqs, verbose = verbose)
  if (browse==TRUE){
    BrowseSeqs(seqs.aligned)
    }
  return(seqs.aligned)
}
# Alignment
polynoids.align <- do_alignment(polynoids.sample, polynoids.sample$seqlabel)
polynoids.bin <- as.DNAbin(polynoids.align)
rm(do_alignment)
```


Only Polynoidae specimens having complete taxonomic information and *cytochrome c oxidase subunit I* (CO1) DNA sequences of appropriate length (630-680 bp) with no more than 2% ambiguous bases were retained. This filtered dataset has `r nrow(polynoids.df)` specimens from `r length(unique(polynoids.df$genus_name))` genera, with `r length(unique(polynoids.df$species_name))` species in total. A single representative from each species was chosen at random for alignment and  phylogenetic analysis. Sequence properties (length, GC-content) and species distribution among genera were examined during data selection (*eg*. fig. 1). A multiple sequence alignment was performed and checked after sequence orientation.

```{r EDA, echo=FALSE, fig.height=2, fig.width=6, fig.align='center'}
coi.sample.eda
```
__Figure 1__. Exploratory analysis of sequence length (l), GC composition (c), and species distribution among genera (r) in the polynoid *cytochrome c oxidase subunit 1* sequences dataset used in this analysis (with one sequence per species). 


```{r include=FALSE}
rm(polynoids.bold, polynoids.df, trim_and_filter_seqs, EDA_plot, coi.sample.eda, coi.eda.filtered, coi.eda.all)
```

-----

### 4. Tools for model tests and phylogenetic analysis

This work relied heavily on the `phangorn` package for maximum-likelihood phylogenetic inference and model tests [@Schliep2011]. Many R packages are available for ML-based phylogenetics (*eg*. RAxML), however `phangorn` has a large selection of models of molecular evolution (I used JC, F81, K80, HKY, SYM, GTR, though there are more) and statistical tests that were well-suited to this analysis; furthermore, any of the substitution models can be paired with +G (gamma-distributed substitution rate variance across sites) or +I (proportion of invariant sites) models, which was not the case with some other packages (eg. `DECIPHER` has no +I and needs to get with the times). Sequence alignment was done with the package `DECIPHER` [@Wright2016], as well as an alternate version of ML phylogenetic clustering for comparison with the best model from `phangorn`. The `ape` package [@Paradis2012]  was used to compute some sequence distances; `dendextend` was used for dendrograms and tanglegrams. For the most part, all these analyses are well-demonstrated in the package vignettes and documentation (see references); this work extends those examples with pairwise statistical tests and tree distance tests for all models, with heatmaps for visualization (`ggplot2`).

-----

### 5. Model testing and phylogenetic Analysis

Molecular evolution model testing with `phangorn::modelTest`

```{r phangorn model test, message=FALSE}
# need phyDat format for test
polynoids.phy <- as.phyDat(polynoids.bin)
# calculate dist, make basic tree, then test subs. models
polynoids.dist <- phangorn::dist.ml(polynoids.phy, model='JC69')
polynoids.nj <- NJ(polynoids.dist)
# test default models selection; trace 0 makes it silent
model_test <- phangorn::modelTest(
  polynoids.phy, polynoids.nj, control = pml.control(trace=0)
  )
rm(polynoids.phy, polynoids.dist, polynoids.nj)
```


```{r get models and do pairwise tests}
get_all_models <- function(model_test_obj){
  # get models from phangorn modelTest output using names list
  extract_model <- function(name, env){ 
    return(eval(get(name, env), env=env)) # retrieves model from model test output
  }
  tests.env <- attr(model_test_obj, "env")
  model.names <- ls(env=tests.env)[2:25]
  models <- lapply(model.names, extract_model, tests.env)
  names(models) <- model.names
  return(models)
}
get_models_pw_df <- function(models){ 
  # get_models_pw_df makes long-form df of pairwise matrix of models#
  m.names <- names(models)
  m.len <- length(m.names)
  # create long table of models, pairwise
  models.df <- tibble(
    name1 = rep(m.names, each = m.len),
    name2 = rep(m.names, times = m.len),
    model1 = rep(models, each = m.len),
    model2 = rep(models, times = m.len),
  ) %>%
    # no self-comparison; (make diagonals blanks)
    filter(!name1==name2) %>%
    rowwise() %>%
    mutate(tree1 = list(pluck(model1, 'tree')), 
           tree2 = list(pluck(model2, 'tree'))) %>%
    # compute tree dists: Robinson-Foulds, weightedRF, branch-score diff, path diff
    mutate(
      RF.dist = phangorn::RF.dist(tree1, tree2, check.labels = TRUE),
      wRF.dist = phangorn::wRF.dist(tree1, tree2, check.labels = TRUE),
      KF.dist = phangorn::KF.dist(tree1, tree2, check.labels = TRUE),
      path.dist = phangorn::path.dist(tree1, tree2, check.labels = TRUE)
    ) %>%
    ungroup()
  return(models.df)
}
# get models, make a df, join model scores
models <- get_all_models(model_test)
models.df <- enframe(models) %>%
  left_join(., model_test, by = c('name' = 'Model')) 
# just want names and AICc scores
AICc_scores <- model_test %>% select(Model, AICc)

# get all the models and perform treedist tests; calc deltaAIC
models.pairwise.df <- get_models_pw_df(models) %>%
  left_join(AICc_scores, by = c('name1' = 'Model')) %>%
  dplyr::rename(AICc.1 = AICc) %>%
  left_join(AICc_scores, by = c('name2' = 'Model')) %>%
  dplyr::rename(AICc.2 = AICc) %>%
  mutate(deltaAICc = AICc.2 - AICc.1)
```

Single log-likelihood ratio test for HKY+G+I & GTR+G+I
```{r}
# grab best model from test, HKY+G+I
HKY_IG.phyl <- models.pairwise.df %>% 
  filter(name1 == "HKY+G+I") %>% 
  pull(model1) %>% pluck(1)
# grab most parameter-rich model, GTR+G+I
GTR_IG.phyl <- models.pairwise.df %>% 
  filter(name1 == "GTR+G+I") %>% 
  pull(model1) %>% pluck(1)
# do log-likelihood ratio test -> not significant!
LRT <- anova(HKY_IG.phyl, GTR_IG.phyl)
# check topological dist -> the only difference is branch lengths!
treedists <- treedist(HKY_IG.phyl$tree, GTR_IG.phyl$tree)
```

-----

### 6. Figures

<!-- 6. Quality of Visualizations (20%) -->
<!-- Throughout, ensure that your figures are clear and well labeled. Even for simple figures, such as histograms, ensure that you have accurate, informative axis labels. Also, consider readability, visual appeal, and accessibility. Use well-differentiated colours, and avoid relying upon the red-green spectrum to convey scientifically important information. Remember, you can consider using a combination of colour and symbol/pattern to convey your meaning. The grade in this section is based upon quality and novelty, not having the maximum permissible number of figures. You should have a total of 4-6 figures for your project (excluding the bonus section, should you choose to complete that section). -->


```{r model test plot, fig.align='center', fig.height=3.5, fig.width=4}
# plotted in fig2: logLik, AICc, and BIC; model deg.freedom
model_test_plot <- model_test %>%
  select(Model, AICc, BIC, logLik, df)  %>%
  pivot_longer(AICc:df) %>%
  mutate(name = as_factor(name)) %>%
  ggplot(aes(x=Model, y=value, colour = name)) + 
  geom_point() + 
  labs(x = '', y = '') +
  facet_grid(name~., scales = 'free_y') + 
  scale_color_carto_d() + theme_light() +
  theme(legend.position = 'none', 
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 7)
  )
model_test_plot
```
__Figure 2__. Testing of all DNA substitution models (JC, F81, K80, HKY, TrN, TPM, K81, TIM, TVM, SYM, GTR) with and without added site-specificity (G, I) for the polynoid CO1 sequences, using `phangorn::modelTest`. The AIC, BIC and log likelihood scores (y-axis) are shown for each model (x-axis), where smaller AICc and BIC scores and larger log likelihood scores indicate better fit. The degrees of freedom for each model are included (df). 

```{r, fig.align = 'center', fig.width=8, fig.height=4}
# Create heatmap for log-likelihood difference
a <- ggplot(models.pairwise.df, aes(y=name2, x=name1, fill=deltaAICc)) +
  geom_tile(na.rm = TRUE) +
  scale_fill_viridis_c(option='D',direction = -1) +
  theme_light() +
  labs(x='', y='', 
       fill='AICc(y) - AICc(x)') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = 'left')
# Create heatmap for weighted Robinson-Foulds distance
b <- models.pairwise.df %>%
  # keep only 'upper triangle' if df was sq matrix
  mutate(index1 = match(name1, names(models)),
         index2 = match(name2, names(models))) %>%
  filter(!index2 <= index1) %>%
  select(-index1, -index2) %>%
  ggplot(aes(y=name2, x=name1, fill=wRF.dist)) +
  geom_tile(na.rm = TRUE) +
  scale_fill_viridis_c(option='C') +
  theme_light() +
  labs(x='', y='',
       fill='wieghted RF distance') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = 'top')
(a+b)

```
__Figure 3__. Heatamaps for the difference in AIC with small-samples correction (AICc) between each pair of models (left), and weighted Robinson-Foulds topological distance from pairwise comparisons all models tested using `phangorn` (right). Note that smaller AICc scores indicate better fit, so rows that have smaller delta(AICc) values represent better models (and the opposite is true for columns), where deltaAICc = AICc(model.y) - AICc(model.x).


```{r fig.height=6, fig.width=4}
library(phylogram)
# need `phylogram` library to convert class phylo to dendrogram
HKY_dendrogram <- as.dendrogram(HKY_IG.phyl$tree) %>%
  set('labels_cex', 0.4) %>%
  color_branches(k=38) 
ggplot(HKY_dendrogram, horiz = TRUE)

```
  
  __Figure 4.__ Dendrogram of the maximum-likelihood phylogenetic estimation with the HKY+G+I model.

### 7. Supplemental 

```{r supPlot1, echo = FALSE, fig.height=2.5, fig.width=5.5, fig.align='center'}
distVector <- function(model, dnabin){
  # return numeric vector of the upper triangle of the distance matrix
  dist <- dist.dna(x = dnabin,
                   model = model,
                   as.matrix = FALSE,
                   pairwise.deletion = TRUE)
  return(as.numeric(dist))
}
# Molecular evolution analysis ----
# create a df with all the distance measures as columns
dist.df <- tibble(model = c('raw','jc69','N', 'k80', 'tn93', 'gg95', 'TS', 'TV')) %>%
  mutate(data = lapply(model, function(x) distVector(x, polynoids.bin))) %>%
  pivot_wider(names_from = 'model', values_from = 'data') %>%
  unnest(cols = raw:TV)

# plot p-distance ~ k80 distance
raw_k80.plot <- dist.df %>%
  ggplot(aes(x=k80, y=raw)) + 
  geom_point(alpha = 0.5, size = 0.5, pch = 1) +
  # add diagonal line
  geom_path(
    data = data.frame(x = seq(0, 1, 0.05), y = seq(0, 1, 0.05), group = factor('group')),
    aes(x = x, y = y, group = group), alpha = 0.7, lty = 2) +
  theme_classic() +
  xlim(c(0, 0.4)) + ylim(c(0, 0.4)) +
  labs(x = 'K2P distance', y = 'p-distance (raw)')

# plot transitions and transversions ~ k80 distance
ts_tv.plot <- dist.df %>%
  select(k80, TS, TV) %>%
  # manipulate table into long-form
  pivot_longer(cols=c(TS, TV),
               names_to = 'Substitution', values_to = 'Proportion') %>%
  # create scatterplot w smooth fits
  ggplot(aes(x = k80, y = Proportion, color = Substitution, group = Substitution)) +
  geom_point(size=0.04, alpha = 0.3) +
  geom_smooth(method = 'loess', formula = 'y~x', se=FALSE) +
  theme_classic() +
  scale_color_carto_d('Type') +
  guides(colour = guide_legend(override.aes = list(size=5, alpha=1, pch=15))) +
  labs(x = 'K2P distance', y = 'Substiutions')
raw_k80.plot + ts_tv.plot
rm(raw_k80.plot, ts_tv.plot, dist.df)
```
__Figure 4__. Saturation plots: p-distance underestimates real evolutionary distance as distance increases, while the Kimura 2-parameter is more robust to the saturation of transitions (right). The proportion of transitions to is generally greater that transversions amongst the family Polyploidae, where phylogenetic distances are relatively small and transitions are not entirely saturated; if we were to compare all polychaete species, for example, the lines would cross and transversions would be more common as transitions become saturated. 

-----

DECIPHER has no implementation of the GTR+I+G model, so I wondered how it's chosen model would compare. TN93+G(4) was chosen and the scores were very similar to the HKY+G+I model generated in the `phangorn` model testing. I also wanted to see how different the phylogenies would be compared a distance-based method with the same substitution model. 

```{r ML clustering with decipher, message=FALSE}
## Maximum-likelihood hierarchical clustering
clusters.ML <- DECIPHER::IdClusters(
  myXStringSet = polynoids.align,
  myDistMatrix = DistanceMatrix(polynoids.align, type='dist'),
  method = 'ML',  cutoff = 0.2, showPlot = FALSE,
  type = "both", verbose = FALSE)
clusters.Dist <- DECIPHER::IdClusters(
  myXStringSet = polynoids.align,
  myDistMatrix = DistanceMatrix(polynoids.align, type='dist'),
  model = 'TN93', method = 'NJ', cutoff = 0.2, showPlot = FALSE,
  type = "both", verbose = FALSE)
```

```{r tanglegram, fig.height=6, fig.width=7}
# library(phylogram)
dend.list <- dendlist(
    clusters.Dist[[2]] %>% set('labels_cex', 0.65) %>%
      set("branches_lwd", 1),
    clusters.ML[[2]] %>% set('labels_cex', 0.65) %>%
      set("branches_lwd", 1)
    ) %>%
  dendextend::untangle() %>% 
# Make tanglegram
  tanglegram(
    match_order_by_labels = TRUE,
    just_one = TRUE,
    main_right = 'ML, TN93+G model',
    main_left = 'dist-based, TN93 model',
    columns_width = c(6, 1, 6),
    left_dendo_mar = c(2, 2, 2, 6),
    right_dendo_mar = c(2, 6, 2, 2)
  )
```


### 8. Results and Discussion


There is disagreement in the literature as to whether model selection significantly impacts maximum likelihood (ML) phylogenetic inferences@Abadi2019; @Gerth2019]. In this project, I wanted to see if model selection was in fact a crucial step in an ML phylogenetic analysis of the family Polynoidae, using CO1 DNA sequences. In model testing HKY+I+G was the best performing model across all key criteria (AICc, BIC, log-likelihood; figs. 2, 3-left); however, differences in goodness-of-fit between this model and the most parameter-rich, GTR+G+I, were not significant in a likelihood ratio test. In fact, the inclusion of the variation among sites (+G), and static sites (+I) into the model had much more impact that differences between substitution models (fig.3-left). Further, the model choice did not affect the partitioning of polynoid species whatsoever (according to the Robinson-Folds RF distance), though when branch lengths were accounted for (in weighted RF), the HKY+G, HKY+G+I and GTR+G+I models were slightly different to all other models tested (fig. 3-right). Thus, my results indicated that most parameter-rich models will produce very similar phylogenies and that model testing was unnecessary. This was not unexpected, as Abadi *et al.* (2019) have suggested that ML model testing generally does not greatly impact inferences, in contravention with *eg*. Posada *et al.* (2008). 

Avoiding model testing is most beneficial when datasets are large and optimization of many models would be time-consuming and computationally intensive. Additionally, simpler models would generally only be preferable to the more realistic GTR+I+G when the data are limited and error on parameter estimates large. While my results agree that model testing generally seems unnecessary, this project only used a single, small dataset (n=95 species), of a single region (CO1), and only for DNA sequences (fig. 4); this dataset was drawn from a single speciose family where distances are relatively small (tree height ~ 0.2, depending on model) and was chosen mostly for convenience and tractability. As such, it certainly cannot be said that the results will generalize to all other datasets. Furthermore, some available models were not tested at all for brevity (TIM, TVM) and only a single algorithm, `phangorn::modelTest`[@Schliep2011], was used for testing, though many other similar tools could have been employed for comparison. Perhaps inclusion of all models or other tools could have uncovered a model with significantly better fit or producing different topology. To determine whether model choice impacts phylogenetic inferences in general would require analysis of multiple, varied datasets and this is generally done through simulated data with known topology to evaluate concordance [@Abadi2019]. As such, my results should only be interpreted in the limited context of the Polyploidae CO1 DNA sequences analyzed and the selection of models that were tested. 

Model testing is but one step in phylogenetic analysis, albeit a particularly time-consuming one. Many other choices could have a far greater impact on inferences, for example: the selection of phylogenetic marker(s); use of DNA sequences versus their more conserved translations; inclusion of structural information, in the case of rRNAs; alignment parameters, etc. A more complete version of my analysis would have compared the utility of model selection in these various scenarios; this could provide further support to the recent proposition that model selection is unimportant for ML phylogeny estimation [@Abadi2019], or it's rebuttal [@Gerth2019]. This could be particularly helpful in the context of analysis of high-throughput sequencing data, where model selection is a major burden to the speed of the analysis.

Conducting this analysis gave me some extra insight into the details that should be considered when applying ML phylogeny methods to real biological data. While the scope of this project was fairly limited, I gained exposure to many different tools aside from those applied here. Phylogenetic analysis in the R environment was a bit daunting when starting out; though over time, I developed a grasp of the statistical underpinnings of maximum likelihood methods, how a workflow could be structured, which tools are commonly applied, how to evaluate different models, and which additional packages are available for specific tasks such as visualization. This will undoubtedly be of benefit in the future.

----

### 9. Acknowledgements

I would like to thank Dr. Sarah Adamowicz and Jacqueline May for the advice and support they provided in developing ideas for this analysis and suggestions about which packages and tools might be useful. 

----

### 10. References

In preparing this work, I found the following tutorials, vignettes, and documentation most helpful:

  - [Estimating phylogenetic trees with phangorn, K.P. Schliep](https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf)  
  - [Phagorn specials vignette, K.P. Schliep](https://cran.r-project.org/web/packages/phangorn/vignettes/phangorn-specials.pdf)   
  - [Getting the models from the output of `phangorn::modelTest`](https://rdrr.io/cran/phangorn/man/modelTest.html)  
  - [The art of multiple sequence alignment in R by E.S. Wright](https://www.bioconductor.org/packages/release/bioc/vignettes/DECIPHER/inst/doc/ArtOfAlignmentInR.pdf)  
  - [Introduction to `dendextend` by T. Galili](https://cran.r-project.org/web/packages/dendextend/vignettes/dendextend.html)  
  - [Genetic data analysis using R: introduction to phylogenetics by T. Jombart](http://adegenet.r-forge.r-project.org/files/MRC-session2-tuto.1.3.pdf)

----

#### Books and Journal Articles
  
*NB: Any articles not available through the University of Guelph subscriptions were shamelessly pirated through [sci-hub.tw](https://scihubtw.tw/)*


  
  
  
