---
title: "Title..."
author: "Jason Moggridge"
date: "28/10/2020"
output: pdf_document

---
<!-- other yaml matters -->
<!-- bibliography: references.bib -->
<!-- highlight: kate -->
<!-- urlcolor: blue -->

<!-- 
Paragraph 1: What is the overarching topic you are interested in? Why is this scientifically important and interesting and/or societally relevant?o

Paragraph 2: You can outline what is an important sub-area of research or a gap in knowledge. (As an example, perhaps your broader theme is gene expression analysis, and in paragraph 2 you could narrow in to introduce the importance of statistical choices.)o
Paragraph 3: What is the specific objective of your study? What kind of study are you performing? What will you d 
-->

### Introduction

Exploring the impact of models of molecular evolution on inferred distance
Polychaeta are diverse group... many cryptic species..


### Dataset


```{r setup}
knitr::opts_chunk$set(
  cache = TRUE,
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
options(tinytex.verbose = TRUE)
```


```{r}
# Libraries
library(tidyverse)
library(bold)
library(Biostrings)
library(DECIPHER)
# library(ape)
library(beepr)
library(rcartocolor)
library(ggthemes)
library(patchwork)
```


Downloading barcoding data from BOLD (change code chunk header to `eval=TRUE` to download data)
```{r eval=FALSE}
 # BOLD database results for taxon 'Polychaeta'; downloaded on 2020/10/31.
polychaeta.bold <- bold::bold_seqspec(taxon = 'Polychaeta')
write_rds(polychaeta.bold, './data/polychaeta.bold.rds', compress = 'gz')
```



Tidy up C01 and CytB sequences
```{r filter}
filter_seqs <- function(bold.df, markers, minlen, maxlen, n_threshold){
  ## Filter data and tidy up COI sequences
  bold.df <- bold.df %>%
    # replace blanks with na's and make strings factors, except for sequences
    mutate(across(where(is.character), ~na_if(.x, ''))) %>%
    mutate(across(where(is.character), as.factor)) %>%
    mutate(nucleotides = as.character(nucleotides))  %>%
    # Keep only specimens with desired marker gene sequences
    filter(!is.na(nucleotides) & markercode %in% markers) %>%
    # Remove gaps and outer N's
    mutate(nucleotides.trim = str_remove_all(nucleotides, '\\s|-|^N+|N+$')) %>%
    mutate(slen = nchar(nucleotides.trim)) %>%
    # filter by length and any with too many Ns
    filter(slen > minlen & slen < maxlen & 
             !str_count(nucleotides.trim, 'N')/slen > n_threshold)
  return(bold.df)  
}

### Tidy BOLD data
# keep only specimens with taxonomy info to species-rank
polychaeta.raw <- read_rds('./data/polychaeta.bold.rds') %>%
  filter(!is.na(species_name))
dim(polychaeta.raw)
# tidy CO1 and CytB data
polychaeta.CO1 <- filter_seqs(polychaeta.raw, 'COI-5P', 600, 700, 0.01)
# polychaeta.CytB <- filter_seqs(polychaeta.raw, 'CYTB', 1100, 1200, 0.01)
# species_CO1_CytB <- inner_join(polychaeta.CO1, polychaeta.CytB, by = "species_name")
dim(polychaeta.CO1)
# dim(polychaeta.CytB)
rm(polychaeta.raw)
```


There are too many sequences to align them all, so we will select a random sequence to represent each genus. After sampling, some sequences were found to

```{r}
# some sequences that did not align well, could be erroneous
poorly_aligned <- c(
  'GBAN1575-07', 'GBAN1576-07', 'GBAN2779-10', 'GBAN6823-15', 'GBAN13543-19', 'GBAN13546-19', 'GBAN13547-19', 'GBAN13548-19', 'GBAN13550-19', 'GBAN13551-19', 'GBAN13552-19', 'GBAN13553-19', 'GBMNB34178-20', 'GBMNB34179-20', 'GBMNB34180-20', 'GBMNB34182-20', 'GBMNB34183-20', 'CCFZP144-19', 'HZPLY376-13'
)
# they are mainly from two families
poorly_aligned <- polychaeta.CO1 %>%
  filter(processid %in% poorly_aligned) %>%
  mutate(family_name = fct_drop(family_name))
summary(poorly_aligned$family_name)

# Prepare down-sampled dataframe for alignment
set.seed(1)
polychaeta.CO1.sample <- polychaeta.CO1 %>%
  # filter(family_name == 'Polynoidae') %>%
  # take 1 sequence per genus
  filter(!is.na(genus_name)) %>%
  group_by(genus_name) %>%
  sample_n(1) %>%
  # exclude sequences known to align poorly
  filter(!processid %in% poorly_aligned$processid) %>%
  data.frame() %>%
  # create labels for the alignment
  mutate(name = paste(species_name, processid)) %>%
  # tidy up factors and remove unneeded cols
  mutate(across(where(is.factor), fct_drop)) %>%
  select(processid, name, nucleotides.trim, slen, 
         family_name, genus_name, species_name)
#
dim(polychaeta.CO1.sample)
rm(poorly_aligned, polychaeta.CO1)


```


Alignments and distance matrices
```{r}
# Make 'seq' a DNA stringset
polychaeta.CO1.sample$seq <-
  Biostrings::DNAStringSet(polychaeta.CO1.sample$nucleotides.trim)
# Attach labels with taxonomy + id
names(polychaeta.CO1.sample$seq) <- polychaeta.CO1.sample$name

# Ensure sequences are similarly oriented
polychaeta.seqs$seq <- OrientNucleotides(polychaeta.seqs$seq)

# Alignment
polychaeta.align <- DECIPHER::AlignSeqs(polychaeta.seqs$seq, verbose = TRUE)
```


Distance Matrices
```{r}

```

Hierarchical Clustering
```{r}

```

Dendrograms
```{r}

```


Comparing rates of substitution between genes

```{r}
distVector <- function(model, dnabin){
  # return numeric vector of the upper triangle of the distance matrix
  dist <- dist.dna(x = dnabin,
                   model = model,
                   as.matrix = FALSE,
                   pairwise.deletion = TRUE)
  return(as.numeric(dist))
}
```


```{r}

```

